morpho-checkpoint-node/src/Morpho/Tracing/Pretty.hs

```diff
@@ -349,37 +707,143 @@ instance (Show tx, Show txid) => MPretty (TraceTxSubmissionOutbound txid tx) whe
+    "Trying to connect to" <+> mpretty addr
+  mpretty (SubscriptionTraceSkippingPeer addr) =
+    "Skipping peer" <+> mpretty addr
+  mpretty SubscriptionTraceSubscriptionRunning =
+    "Required subscriptions started"
+  mpretty (SubscriptionTraceSubscriptionWaiting d) =
+    "Waiting on" <+> pretty d <+> "active connections"
+  mpretty SubscriptionTraceSubscriptionFailed =
+    "Failed to start all required subscriptions"
+  mpretty (SubscriptionTraceSubscriptionWaitingNewConnection delay) =
+    "Waiting" <+> viaShow delay <+> "before attempting a new connection"
+  mpretty (SubscriptionTraceStart val) =
+    "Starting Subscription Worker, valency" <+> pretty val
+  mpretty (SubscriptionTraceRestart duration desiredVal currentVal) =
+    "Restarting Subscription after"
+      <+> viaShow duration
+      <+> "desired valency"
+      <+> pretty desiredVal
+      <+> "current valency"
+      <+> pretty currentVal
+  mpretty (SubscriptionTraceConnectionExist dst) =
+    "Connection Existed to" <+> mpretty dst
+  mpretty (SubscriptionTraceUnsupportedRemoteAddr dst) =
+    "Unsupported remote target address" <+> mpretty dst
+  mpretty SubscriptionTraceMissingLocalAddress =
+    "Missing local address"
+  mpretty (SubscriptionTraceAllocateSocket addr) =
+    "Allocate socket to" <+> mpretty addr
+  mpretty (SubscriptionTraceCloseSocket addr) =
+    "Closed socket to" <+> mpretty addr
+
+instance MPretty ConnectResult where
+  mpretty ConnectSuccess = "Successful connection"
+  mpretty ConnectSuccessLast = "Successfully connection, reached the valency target. Other ongoing connection attempts will be killed."
+  mpretty ConnectValencyExceeded = "Someone else manged to create the final connection to a target before us."
+
+instance MPretty DnsTrace where
+  -- DnsTrace already has a nice humain readable Show instance
   mpretty = viaShow
 
-instance MPretty (WithIPList (SubscriptionTrace Socket.SockAddr)) where
+instance MPretty Socket.SockAddr where
   mpretty = viaShow
 
-instance (Show peer) => MPretty (WithMuxBearer peer MuxTrace) where
+instance (Show a) => MPretty (NtN.ConnectionId a) where
   mpretty = viaShow
 
+instance MPretty a => MPretty (WithIPList a) where
+  mpretty (WithIPList (NtN.LocalAddresses Nothing (Just ipv6) Nothing) wilDsts wilEvent) =
+    "IPs:" <+> viaShow ipv6 <+> viaShow wilDsts <+> mpretty wilEvent
+  mpretty (WithIPList (NtN.LocalAddresses (Just ipv4) Nothing Nothing) wilDsts wilEvent) =
+    "IPs:" <+> viaShow ipv4 <+> viaShow wilDsts <+> mpretty wilEvent
+  mpretty (WithIPList (NtN.LocalAddresses Nothing Nothing (Just unix)) wilDsts wilEvent) =
+    "IPs:" <+> viaShow unix <+> viaShow wilDsts <+> mpretty wilEvent
+  mpretty (WithIPList (NtN.LocalAddresses (Just ipv4) (Just ipv6) Nothing) wilDsts wilEvent) =
+    "IPs:" <+> viaShow ipv4 <+> viaShow ipv6 <+> viaShow wilDsts <+> mpretty wilEvent
+  mpretty WithIPList {wilSrc, wilDsts, wilEvent} =
+    "IPs:" <+> viaShow wilSrc <+> viaShow wilDsts <+> mpretty wilEvent
```

Comment:
This instance can be improved. I'm thinking

> src: [\<ip1\>, \<ip2\>], dst: [\<ip1\>, \<ip2\>], event: \<event\>

You can turn a `LocalAddresses addr` into `[addr]` with
```haskell
(LocalAddresses a b c) = catMaybes [a, b, c]
```

then you only need a single pattern match